File: bkasm/src/parser.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include "bkasm.h"
#include "parser.h"
#include "asmast.h"
#include "errors.h"
#include "inbuf.h"
#include "mathexpr.h"
#include "nodes.h"

static void parse_comment(Parser *self, Lexer *lexer);
static Node *parse_op(Parser *self, Lexer *lexer);
static Expr *parse_param(Parser *self, Lexer *lexer);
static void parse_statement(Parser *self, Lexer *lexer);

static Node *parse_var(Parser *self, Lexer *lexer)
{
    Node *node;

    Lexema m_token = lexer->token;
    char *l_ident = m_token.ident;
    lexer->nextTok(lexer);
    m_token = lexer->token;
    node = NULL;

    if (m_token.type == TOK_COLON)
    {
        node = node_create_label(l_ident);
        lexer->skipOne(lexer);
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        m_token = lexer->token;
        if (m_token.type == TOK_EQU)
        {
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            node->label.target = parse_param(self, lexer);
            node->label.target_type = TOK_IDENT;
            lexer->skipUntil(lexer, 10);
            parse_comment(self, lexer);
            // pass node throw standart codeflow
        }
        else
        {
            node->label.target = register_create(TOK_REGPC, "PC");
            node->label.target_type = TOK_REGPC;
            ast_add_statement(node, self->ast); // add standart label
            parse_statement(self, lexer);       // parse remain part of string
            node = NULL;                        // nothing to return, all parts were parsed
        }
        free(l_ident);
    }
    else
    {
        throw_error(E_UNKIDENT, l_ident);
        free(l_ident);
        lexer->skipUntil(lexer, 10);
        lexer->skipOne(lexer);
    }
    return node;
}

static void parse_comment(Parser *self, Lexer *lexer)
{
    Lexema m_token = lexer->token;
    InbufCurrentString *currstr = inbuf_currstr();
    switch (m_token.kind)
    {
    case SYM:
        switch (m_token.type)
        {
        case L_EOL:
            lexer->skipOne(lexer);
            break;
        case L_EOF:
            lexer->skipOne(lexer);
            return;
        case TOK_SEMICOLON:
            lexer->skipUntil(lexer, 10);
            lexer->skipOne(lexer);
            break;
        default:
            fprintf(stderr, "In string: %d %s\n", currstr->num, currstr->str);
            throw_error(E_UNEXPSYM, m_token.ident);
            lexer->skipUntil(lexer, 10);
        }
        break;
    default:
        fprintf(stderr, "In string:  %d %s\n", currstr->num, currstr->str);
        throw_error(E_UNEXPSYM, m_token.ident);
        lexer->skipUntil(lexer, 10);
        break;
    }
}

static Expr *parse_term(Parser *self, Lexer *lexer)
{
    Lexema m_token = lexer->token;
    Expr *result = NULL;
    char *m_ident = m_token.ident;
    switch (m_token.type)
    {
    case TOK_NUM:
        result = const_create(m_token.value);
        break;
    case TOK_IDENT:
        result = var_create(m_token.ident);
        free(m_ident);
        break;
    default:
        throw_error(E_UNEXPTOKEN, m_token.ident);
        result = const_create(0);
    }
    return result;
}

static Expr *parse_multiplication(Parser *self, Lexer *lexer)
{

    Lexema m_token;
    Expr *result = parse_term(self, lexer);
    lexer->skipWhile(lexer, ' ');
    lexer->nextTok(lexer);
    m_token = lexer->token;
    for (; m_token.type == TOK_ASTERISK;)
    {
        Expr *expr = math_create_multiplication(TOK_ASTERISK);
        expr->op.setlparam(expr, result);
        lexer->skipOne(lexer);
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        ;
        expr->op.setrparam(expr, parse_term(self, lexer));
        result = expr;
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        m_token = lexer->token;
    }
    return result;
}

static Expr *parse_addition(Parser *self, Lexer *lexer)
{
    Lexema m_token;
    Expr *result = parse_multiplication(self, lexer);
    lexer->skipWhile(lexer, ' ');
    lexer->nextTok(lexer);
    m_token = lexer->token;
    for (; m_token.type == TOK_PLUS || m_token.type == TOK_MINUS;)
    {
        if (m_token.type == TOK_MINUS)
        {
            Expr *expr = math_create_addition(TOK_MINUS);
            expr->op.setlparam(expr, result);
            lexer->skipOne(lexer);
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            expr->op.setrparam(expr, parse_multiplication(self, lexer));
            result = expr;
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            m_token = lexer->token;
        }
        if (m_token.type == TOK_PLUS)
        {
            Expr *expr = math_create_addition(TOK_PLUS);
            expr->op.setlparam(expr, result);
            lexer->skipOne(lexer);
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            expr->op.setrparam(expr, parse_multiplication(self, lexer));
            result = expr;
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            m_token = lexer->token;
        }
    }
    return result;
}

static Expr *parse_param(Parser *self, Lexer *lexer)
{
    Expr *expr = NULL;
    Lexema m_token = lexer->token;
    switch (m_token.kind)
    {
    case REG:
        expr = register_create(m_token.type, m_token.ident);
        break;
    case VAR:
    case CONST:
        switch (m_token.type)
        {
        case TOK_IDENT:
        case TOK_NUM:
            expr = parse_addition(self, lexer);
            break;
        default:
            throw_error(E_UNKIDENT, m_token.ident);
        }
        break;
    default:
        throw_error(E_UNEXPTOKEN, m_token.ident);
        exit_nicely(E_UNEXPTOKEN);
    }
    lexer->skipWhile(lexer, ' ');
    return expr;
}

static Node *parse_op(Parser *self, Lexer *lexer)
{
    Instruction *op;
    Lexema op_token = lexer->token;
    Node *node = node_create_instruction(op_token.ident, op_token.type);
    InbufCurrentString *currstr = inbuf_currstr();

    switch (op_token.type)
    {
    // multibyte arrays
    case TOK_DB:
        op = (Instruction *)node;
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        op->lparam = parse_param(self, lexer);
        ast_add_statement(node, self->ast);
        lexer->skipWhile(lexer, ',');
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        for (; lexer->token.kind == CONST;)
        {
            node = node_create_instruction(op_token.ident, op_token.type);
            op = (Instruction *)node;
            op->lparam = parse_param(self, lexer);
            ast_add_statement(node, self->ast);
            lexer->skipWhile(lexer, ',');
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
        }
        node = NULL;
        break;

    // two opernds mnemonics
    case TOK_LXI:
    case TOK_MOV:
    case TOK_MVI:
        op = (Instruction *)node;
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        op->lparam = parse_param(self, lexer);
        lexer->skipWhile(lexer, ',');
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        op->rparam = parse_param(self, lexer);
        lexer->skipWhile(lexer, ' ');
        break;
    // one operand mnemonics
    case TOK_ACI:
    case TOK_ADC:
    case TOK_ADD:
    case TOK_ADI:
    case TOK_ANA:
    case TOK_ANI:
    case TOK_CALL:
    case TOK_CC:
    case TOK_CM:
    case TOK_CMP:
    case TOK_CNC:
    case TOK_CNZ:
    case TOK_CP:
    case TOK_CPE:
    case TOK_CPI:
    case TOK_CPO:
    case TOK_CZ:
    case TOK_DAD:
    case TOK_DCR:
    case TOK_DCX:
    case TOK_IN:
    case TOK_INR:
    case TOK_INX:
    case TOK_JC:
    case TOK_JM:
    case TOK_JMP:
    case TOK_JNC:
    case TOK_JNZ:
    case TOK_JP:
    case TOK_JPE:
    case TOK_JPO:
    case TOK_JZ:
    case TOK_LDA:
    case TOK_LDAX:
    case TOK_LHLD:
    case TOK_ORA:
    case TOK_ORG:
    case TOK_ORI:
    case TOK_OUT:
    case TOK_POP:
    case TOK_PUSH:
    case TOK_RST:
    case TOK_SBB:
    case TOK_SBI:
    case TOK_SHLD:
    case TOK_STA:
    case TOK_STAX:
    case TOK_SUB:
    case TOK_SUI:
    case TOK_XRA:
    case TOK_XRI:
        op = (Instruction *)node;
        lexer->skipWhile(lexer, ' ');
        lexer->nextTok(lexer);
        op->lparam = parse_param(self, lexer);
        break;
    // NULL operand mnemonics
    case TOK_CMA:
    case TOK_CMC:
    case TOK_DAA:
    case TOK_DI:
    case TOK_EI:
    case TOK_END:
    case TOK_HLT:
    case TOK_NOP:
    case TOK_PCHL:
    case TOK_RAL:
    case TOK_RAR:
    case TOK_RC:
    case TOK_RET:
    case TOK_RIM:
    case TOK_RLC:
    case TOK_RM:
    case TOK_RNC:
    case TOK_RNZ:
    case TOK_RP:
    case TOK_RPE:
    case TOK_RPO:
    case TOK_RRC:
    case TOK_RZ:
    case TOK_SIM:
    case TOK_SPHL:
    case TOK_STC:
    case TOK_XCHG:
    case TOK_XTHL:
        op = (Instruction *)node;
        lexer->skipWhile(lexer, ' ');
        break;
    case TOK_SEMICOLON: // no operation pass-throw comment
        break;
    default:
        fprintf(stderr, "In string: %d %s\n", currstr->num, currstr->str);
        throw_error(E_UNKKEYWORD, op_token.ident);
        lexer->skipUntil(lexer, 10);
        break;
    }
    return node;
}

static void parse_statement(Parser *self, Lexer *lexer)
{
    InbufCurrentString *currstr;
    Lexema m_token = lexer->token;
    // lexer->printTok(lexer->token); // debug
    currstr = inbuf_currstr();

    switch (m_token.kind)
    {
    case VAR:
        switch (m_token.type)
        {
        case TOK_IDENT:
            ast_add_statement(parse_var(self, lexer), self->ast);
            break;
        default:
            fprintf(stderr, "In string: %d %s\n", currstr->num, currstr->str);
            throw_error(E_UNEXPTOKEN, m_token.ident);
        }
        break;
    case SYM:
        switch (m_token.type)
        {
        case TOK_SEMICOLON:
            parse_comment(self, lexer);
            break;
        case L_EOL:
            // printf("< EMPTY STRING >\n");
            lexer->skipOne(lexer);
            break;
        case L_EOF:
            printf(INDENT "< EOF >\n");
            lexer->skipOne(lexer);
            break;
        default:
            fprintf(stderr, "In string: %d %s\n", currstr->num, currstr->str);
            throw_error(E_UNEXPSYM, m_token.ident);
            exit_nicely(E_UNEXPSYM);
        }
        break;
    case OP:
        ast_add_statement(parse_op(self, lexer), self->ast);
        lexer->nextTok(lexer);
        parse_comment(self, lexer);
        break;
    case INT:
    {
        Node *internal = node_create_label("ENDPRG");
        m_token = lexer->token;
        if (m_token.type == TOK_END)
        {
            internal->label.target = register_create(TOK_REGPC, "PC");
            internal->label.target_type = TOK_REGPC;
            ast_add_statement(internal, self->ast); // add standart label
            lexer->skipOne(lexer);
            lexer->skipWhile(lexer, ' ');
            lexer->nextTok(lexer);
            parse_statement(self, lexer); // parse remain part of string
            internal = NULL;              // nothing to return, all parts were parsed
        }
    }
    break;
    case KIND_NONE: // if unexpectd symbol detected
        lexer->skipUntil(lexer, 10);
        lexer->skipOne(lexer);
        break;
    default:
        fprintf(stderr, "In string: %d %s\n", currstr->num, currstr->str);
        throw_error(E_UNEXPTOKEN, m_token.ident);
        assert(0);
        break;
    }
}

void parser_parse(Parser *self, Lexer *lexer)
{
    for (; lexer->nextTok(lexer);)
    {
        parse_statement(self, lexer);
        lexer->skipWhile(lexer, ' ');
    }
}

int parser_init(Parser *parser)
{
    parser->level = 0;
    parser->ast = ast_create();
    ast_init(parser->ast);
    return 1;
}
Parser *parser_create(void)
{
    Parser *m_parser = (Parser *)malloc(sizeof(Parser));
    parser_init(m_parser);
    return m_parser;
}

void parser_free(Parser *_parser)
{
    free(_parser);
}

File: bkasm/csharp/Parser.cs
using System;

namespace Bkasm
{
    public abstract class Parser
    {
        public ASTree ast;
        public eErrorType error;

        public Parser()
        {
            ast = new ASTree();
        }

        protected abstract Node ParseOp(Lexer lexer);

        private Node ParseVar(Lexer lexer)
        {
            var m_token = lexer.token;
            string l_ident = m_token.ident;
            lexer.NextTok();
            m_token = lexer.token;
            Node node = null;

            if (m_token.type == eIdentType.TOK_COLON)
            {
                node = new Node(l_ident);
                lexer.SkipOne();
                lexer.SkipWhile(Constants.SPACE);
                lexer.NextTok();
                m_token = lexer.token;
                if (m_token.type == eIdentType.TOK_EQU)
                {
                    lexer.SkipWhile(Constants.SPACE);
                    lexer.NextTok();
                    node.label.target = ParseParam(lexer);
                    node.label.targetType = eIdentType.TOK_IDENT;
                    lexer.SkipUntil(Constants.EOL_SYM);
                    ParseComment(lexer);
                }
                else
                {
                    node.label.target = new RegExpr((int)eIdentType.TOK_REGPC, "PC");
                    node.label.targetType = eIdentType.TOK_REGPC;
                    ast.AddStatement(node);
                    ParseStatement(lexer);
                    node = null;
                }
            }
            else
            {
                ErrorHandler.ThrowError(eErrorType.E_UNKIDENT, l_ident);
                lexer.SkipUntil(Constants.EOL_SYM);
                lexer.SkipOne();
            }
            return node;
        }

        private void ParseComment(Lexer lexer)
        {
            var m_token = lexer.token;
            var currstr = Inbuf.CurrStr();

            switch (m_token.kind)
            {
                case eIdentKind.SYM:
                    switch (m_token.type)
                    {
                        case eIdentType.L_EOL:
                            lexer.SkipOne();
                            break;
                        case eIdentType.L_EOF:
                            lexer.SkipOne();
                            return;
                        case eIdentType.TOK_SEMICOLON:
                            lexer.SkipUntil(Constants.EOL_SYM);
                            lexer.SkipOne();
                            break;
                        default:
                            error = eErrorType.E_UNEXPSYM;
                            Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                            ErrorHandler.ThrowError(eErrorType.E_UNEXPSYM, m_token.ident ?? "");
                            lexer.SkipUntil(Constants.EOL_SYM);
                            break;
                    }
                    break;
                default:
                    error = eErrorType.E_UNEXPSYM;
                    Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                    ErrorHandler.ThrowError(eErrorType.E_UNEXPSYM, m_token.ident ?? "");
                    lexer.SkipUntil(Constants.EOL_SYM);
                    break;
            }
        }

        private Expr ParseTerm(Lexer lexer)
        {
            var m_token = lexer.token;
            Expr result = null;

            switch (m_token.type)
            {
                case eIdentType.TOK_MINUS:
                    lexer.SkipOne();
                    lexer.SkipWhile(Constants.SPACE);
                    lexer.NextTok();
                    result = ParseTerm(lexer);
                    {
                        var zero = new ConstExpr(0);
                        var expr = new AdditionExpr(eIdentType.TOK_MINUS);
                        expr.lparam = zero;
                        expr.rparam = result;
                        result = expr;
                    }
                    break;
                case eIdentType.TOK_NUM:
                    result = new ConstExpr(m_token.value);
                    break;
                case eIdentType.TOK_IDENT:
                    result = new VarExpr(m_token.ident);
                    break;
                case eIdentType.TOK_LPAREN:
                    lexer.SkipOne();
                    lexer.SkipWhile(Constants.SPACE);
                    lexer.NextTok();
                    result = ParseAddition(lexer);
                    if (lexer.token.type == eIdentType.TOK_RPAREN)
                        lexer.SkipOne();
                    else
                        ErrorHandler.ThrowError(eErrorType.E_SYNTAXERROR, ") expected");
                    break;
                default:
                    ErrorHandler.ThrowError(eErrorType.E_UNEXPTOKEN, m_token.ident ?? "");
                    result = new ConstExpr(0);
                    break;
            }
            return result;
        }

        private Expr ParseMultiplication(Lexer lexer)
        {
            Expr result = ParseTerm(lexer);
            lexer.SkipWhile(Constants.SPACE);
            lexer.NextTok();
            var m_token = lexer.token;

            while (m_token.type == eIdentType.TOK_ASTERISK)
            {
                var expr = new MultiplicationExpr(eIdentType.TOK_ASTERISK);
                expr.lparam = result;
                lexer.SkipOne();
                lexer.SkipWhile(Constants.SPACE);
                lexer.NextTok();
                expr.rparam = ParseTerm(lexer);
                result = expr;
                lexer.SkipWhile(Constants.SPACE);
                lexer.NextTok();
                m_token = lexer.token;
            }
            return result;
        }

        private Expr ParseAddition(Lexer lexer)
        {
            Expr result = ParseMultiplication(lexer);
            lexer.SkipWhile(Constants.SPACE);
            lexer.NextTok();
            var m_token = lexer.token;

            while (m_token.type == eIdentType.TOK_PLUS || m_token.type == eIdentType.TOK_MINUS)
            {
                var expr = new AdditionExpr(m_token.type);
                expr.lparam = result;
                lexer.SkipOne();
                lexer.SkipWhile(Constants.SPACE);
                lexer.NextTok();
                expr.rparam = ParseMultiplication(lexer);
                result = expr;
                lexer.SkipWhile(Constants.SPACE);
                lexer.NextTok();
                m_token = lexer.token;
            }
            return result;
        }

        protected Expr ParseDbParam(Lexer lexer)
        {
            Expr expr = null;
            var m_token = lexer.token;

            if (m_token.kind == eIdentKind.STRING)
            {
                string src = m_token.ident;
                int len = m_token.len;
                int i;

                for (i = 0; i < len - 1; i++)
                {
                    expr = new ConstExpr(src[i]);
                    var node = new Node("DB", eIdentType.TOK_DB, 0);
                    node.op.lparam = expr;
                    ast.AddStatement(node);
                }

                if (len > 0)
                    expr = new ConstExpr(src[len - 1]);
                lexer.SkipOne();
            }
            else
            {
                switch (m_token.kind)
                {
                    case eIdentKind.SYM:
                        if (m_token.type == eIdentType.TOK_MINUS)
                            expr = ParseParam(lexer);
                        else
                            expr = null;
                        break;
                    default:
                        expr = ParseParam(lexer);
                        break;
                }
            }
            return expr;
        }

        protected Expr ParseDwParam(Lexer lexer)
        {
            Expr expr = null;
            var m_token = lexer.token;

            switch (m_token.kind)
            {
                case eIdentKind.SYM:
                    if (m_token.type == eIdentType.TOK_MINUS)
                        expr = ParseParam(lexer);
                    else
                        expr = null;
                    break;
                default:
                    expr = ParseParam(lexer);
                    break;
            }
            return expr;
        }

        protected Expr ParseParam(Lexer lexer)
        {
            Expr expr = null;
            var m_token = lexer.token;

            switch (m_token.kind)
            {
                case eIdentKind.REG:
                    expr = new RegExpr(m_token.value, m_token.ident);
                    break;
                case eIdentKind.VAR:
                case eIdentKind.CONST:
                    switch (m_token.type)
                    {
                        case eIdentType.TOK_IDENT:
                        case eIdentType.TOK_NUM:
                            expr = ParseAddition(lexer);
                            break;
                        default:
                            ErrorHandler.ThrowError(eErrorType.E_UNKIDENT, m_token.ident ?? "");
                            break;
                    }
                    break;
                case eIdentKind.SYM:
                    if (m_token.type == eIdentType.TOK_LPAREN || m_token.type == eIdentType.TOK_MINUS)
                        expr = ParseAddition(lexer);
                    else
                    {
                        ErrorHandler.ThrowError(eErrorType.E_UNEXPTOKEN, m_token.ident ?? "");
                        Globals.ExitNicely(eErrorType.E_UNEXPTOKEN);
                    }
                    break;
                default:
                    ErrorHandler.ThrowError(eErrorType.E_UNEXPTOKEN, m_token.ident ?? "");
                    Globals.ExitNicely(eErrorType.E_UNEXPTOKEN);
                    break;
            }
            lexer.SkipWhile(Constants.SPACE);
            return expr;
        }

        private void ParseStatement(Lexer lexer)
        {
            var currstr = Inbuf.CurrStr();
            var m_token = lexer.token;

            switch (m_token.kind)
            {
                case eIdentKind.VAR:
                    switch (m_token.type)
                    {
                        case eIdentType.TOK_IDENT:
                            ast.AddStatement(ParseVar(lexer));
                            break;
                        default:
                            error = eErrorType.E_UNEXPTOKEN;
                            Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                            ErrorHandler.ThrowError(eErrorType.E_UNEXPTOKEN, m_token.ident ?? "");
                            break;
                    }
                    break;

                case eIdentKind.SYM:
                    switch (m_token.type)
                    {
                        case eIdentType.TOK_SEMICOLON:
                            ParseComment(lexer);
                            break;
                        case eIdentType.L_EOL:
                            lexer.SkipOne();
                            break;
                        case eIdentType.L_EOF:
                            lexer.SkipOne();
                            break;
                        default:
                            error = eErrorType.E_UNEXPSYM;
                            Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                            ErrorHandler.ThrowError(eErrorType.E_UNEXPSYM, m_token.ident ?? "");
                            Globals.ExitNicely(eErrorType.E_UNEXPSYM);
                            break;
                    }
                    break;

                case eIdentKind.OP:
                    ast.AddStatement(ParseOp(lexer));
                    lexer.NextTok();
                    ParseComment(lexer);
                    break;

                case eIdentKind.INT:
                {
                    m_token = lexer.token;
                    if (m_token.type == eIdentType.TOK_INCLUDE)
                    {
                        lexer.SkipWhile(Constants.SPACE);
                        lexer.NextTok();
                        m_token = lexer.token;
                        if (m_token.kind != eIdentKind.STRING)
                        {
                            error = eErrorType.E_SYNTAXERROR;
                            Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                            ErrorHandler.ThrowError(eErrorType.E_SYNTAXERROR, "expected filename after INCLUDE");
                            Globals.ExitNicely(eErrorType.E_SYNTAXERROR);
                        }
                        if (Inbuf.PushFile(m_token.ident) != 0)
                        {
                            error = eErrorType.E_SYNTAXERROR;
                            Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                            Console.Error.WriteLine("Error: Cannot open include file: {0}", m_token.ident);
                            Globals.ExitNicely(eErrorType.E_SYNTAXERROR);
                        }
                        lexer.CurrentChar = (byte)Constants.CH_NULL;
                        break;
                    }
                    {
                        var internal_node = new Node("ENDPRG");
                        if (m_token.type == eIdentType.TOK_END)
                        {
                            internal_node.label.target = new RegExpr((int)eIdentType.TOK_REGPC, "PC");
                            internal_node.label.targetType = eIdentType.TOK_REGPC;
                            ast.AddStatement(internal_node);
                            lexer.SkipOne();
                            lexer.SkipWhile(Constants.SPACE);
                            lexer.NextTok();
                            ParseStatement(lexer);
                            internal_node = null;
                        }
                    }
                    break;
                }

                case eIdentKind.KIND_NONE:
                    lexer.SkipUntil(Constants.EOL_SYM);
                    lexer.SkipOne();
                    break;

                default:
                    error = eErrorType.E_UNEXPTOKEN;
                    Console.Error.WriteLine("In string: {0} {1}", currstr.num, new string(currstr.str));
                    ErrorHandler.ThrowError(eErrorType.E_UNEXPTOKEN, m_token.ident ?? "");
                    break;
            }
        }

        public void Parse(Lexer lexer)
        {
            while (lexer.NextTok() != 0)
            {
                ParseStatement(lexer);
                lexer.SkipWhile(Constants.SPACE);
            }
        }
    }
}
